src/groupdel.8
.\" -*- nroff -*-
.\" Copyright (C) 2003 Thorsten Kukuk
.\" Author: Thorsten Kukuk <kukuk@suse.de>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2 as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software Foundation,
.\" Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.TH groupdel 8 "October 2003" "pwdutils"
.SH NAME
groupdel \- delete a group entry
.SH SYNOPSIS
.TP 8
\fBgroupdel\fR [\fB-D \fIbinddn\fR] [\fB-P \fIpath\fR] [\fB--service \fIservice\fR]
.br
[\fB--help\fR] [\fB-u\fR] [\fB-v\fR] \fIgroup\fR
.SH DESCRIPTION
\fBgroupdel\fR removes a group entry from the local system files or
a LDAP database. A primary group of any existing user is not removeable.
All filesystems must be checked manually to insure that no
files remain with the named group.
.SH OPTIONS
.TP
.BI "\-\-service" " service"
Add the account to a special directory. The default is \fIfiles\fR,
but \fIldap\fR is also valid.
.TP
.BI "\-D, \-\-binddn" " binddn"
Use the Distinguished Name binddn to bind to the LDAP directory.
The user will be prompted for a password for simple authentication.
.TP
.BI "\-P, \-\-path" " path"
The \fIpasswd\fR and \fIshadow\fR files are located below
the specified directory path. \fBchpasswd\fR will use this files,
not \fI/etc/passwd\fR and \fI/etc/shadow\fR.
.TP
.B "\-\-help"
Print a list of valid options with a short description.
.TP
.B "\-u, \-\-usage"
Print a short list of valid options.
.TP
.B "\-v, \-\-version"
Print the version number and exit.
.SH FILES
group \- group information
.SH SEE ALSO
.BR passwd (1),
.BR login.defs (5),
.BR group (5),
.BR groupadd (8),
.BR groupmod (8)
.SH AUTHOR
Thorsten Kukuk <kukuk@suse.de>


NEWS
pwdutils NEWS -- history of user-visible changes.

Copyright (C) 2002-2012 Thorsten Kukuk

Please send bug reports, questions and suggestions to <kukuk@thkukuk.de>.

Version 3.2.19
* Fix passwd -S

Version 3.2.18
* Fix changing of permissions if uid/gid of user will be modified

Version 3.2.17
* Fix useradd.local

Version 3.2.16
* Make sha512 new default password hash

Version 3.2.15
* newgrp: honour dynamically assigned groups

Version 3.2.14
* pwck: check +/- entries, too
* grpck: check */- entries, too
* login.defs: remove LOGIN_RETRIES
 
Version 3.2.13
* Only copy regular files if usermod moves home directory

Version 3.2.12
* Make sure modified group files are written to disk
* Update translations for sv, fr

Version 3.2.11
* Update translations
* Add translation for nl

Version 3.2.10
* Fix encoding of polish translations
* Fix long options of grpck/pwck
 
Version 3.2.9
* Move UMASK from /etc/login.defs to /etc/default/useradd

Version 3.2.8
* Fix usermod manual page
* Add quiet option to pam_rpasswd.so

Version 3.2.7
* Fix possible endless loop in userdel

Version 3.2.6
* Fix rpasswd protocol error
* Update translations

Version 3.2.5
* Add translations for da
* pwconv: Remove temporary files in case of an error

Version 3.2.4
* Make sure changes are really written to disk

Version 3.2.3
* New translations
* Adjust documentation

Version 3.2.2
* Don't lock system accounts by default

Version 3.2.1
* Adjust for new libxcrypt

Version 3.2.0
* Better support for audit plugin
* --preferred-uid/gid option for useradd and groupadd

Version 3.1.4
* Update polish translation.
* Remove FTMP references.
* Better pwconv support for  NIS entries

Version 3.1.3
* Allow compiling without native language support.
* Update swedish translation.

Version 3.1.2
* Check if crontab is installed before calling it.
* Fix seg.fault in pwconv in error case.

Version 3.1.1
* Adjust PAM config files

Version 3.1.0
* gpasswd: Add -l and -u option to lock/unlock group passwords
* passwd: Add support for BioAPI
* Add LDAP support for RFC2307bis

Version 3.0.7
* passwd.1: correct manual page
* passwd: Allow changing passwords with PAM even if service is unknown

Version 3.0.6
* chfn: Fix syntax check for home phone number

Version 3.0.5
* Fix compilation error of PAM module with new Linux-PAM release
* Fix CHARACTER_CLASS in /etc/login.defs

Version 3.0.4
* Fix pam_rpasswd.so

Version 3.0.3
* usermod: Adjust sp_lstchg if -p option is used
* useradd: Preserve existing home directory
* chage: Fix off-by-one errors in output

Version 3.0.2
* Fix loading of logging plugins
* Unify options (use always --service)
* Add -a option to usermod for shadow compatibility

Version 3.0.1
* usermod: Don't remove old home directory if an error occured

Version 3.0
* Add --stdin option for passwd
* Fix useradd -D option
* Fix PAM/syslog interaction
* Fix lot of ignored failing syscalls
* Compile daemons with -fpie/-pie
* Add support for logging plugins
* Add plugin for LAuS support
* Add plugin for audit support
* Add plugin to log with syslog
* Use libnscd
* Add support for GnuTLS
* groupmod: Add options to remove/add single user from a group
* usermod: Add options to remove/add single user from a list of groups
* rpasswd/rpasswdd: Better SLP support

Version 2.6.4
* rpasswd/pam_rpasswd.so: Fix SLP URL parsing

Version 2.6.3
* newgrp: Add -c option to allow sg emulation

Version 2.6.2
* Fix compile problem if SELinux is disabled
* rpasswdd: Re-register SLP service after timeout

Version 2.6.1
* newgrp: Fix NGROUPS_MAX handling

Version 2.6
* Add -f compat flag to groupadd
* Add UTF8/local locale translation
* Make regex for allowed accounts/groups configurable
* Add optional pam_rpasswd module
* Add SLP support to rpasswd/rpasswdd
* Add pwck
* Add grpck
* Add newgrp
* Add /etc/default/passwd support

Version 2.5.1
* Install all manual pages

Version 2.5
* Adjust for libselinux v1.5
* Add pwconv/pwunconv
* Add grpunconv

Version 2.4
* Fix -G handling of usermod
* Fix LDAP support
* Add gpasswd, groupmod and usermod
* Copy always all extended attributes
* Add SELinux support for passwd, chsh and chfn
* Make sure that group, passwd and shadow are always created with
  correct security attributes
* groupadd -g handling fixed
* vipw/vigr added
* Lot of bug fixes
* Add manual pages
* Add groupadd and groupdel
* useradd and userdel can now create/delete LDAP accounts
* Make compile with OpenLDAP 2.0.x again
* Rewrite ldap functions to meet demands
* chpasswd can now change LDAP passwords
* useradd: Lot of new error checks and bug fixes
* chfn/chsh: Allow user to abort with Ctrl-C during data input
* Add test suite

Version 2.3
* Add chpasswd
* Add expiry
* Fix chage options
* rpasswd client now checks server certificate

Version 2.2
* Add "-D binddn" option, to allow modifying LDAP data with a special
  account.
* passwd and chage: Allow modifying of password expiry informations
  stored in a LDAP database.

Version 2.1
* Add --disable-ldap option to configure
* Fix typos in chage.1 manual page
* Rewrite passwd manual page and explain all options

Version 2.0
* Fix -d and -E option of chage
* Use port 774 as new default
* Make compile with glibc 2.3 and gcc 2.95/3.2/3.3

Version 1.99
* Add chage
* Allow modifying of passwd/shadow file not in /etc (only chage yet)

Version 1.98
* Add manual pages for rpasswd and rpasswdd

Version 1.97
* rpasswd now reads config file with server name and port number
* Security fix for rpasswdd: don't allow remote changing of root
  password without authentification
* Add manual page for rpasswd.conf

Version 1.96
* Make compile with gcc 2.x
* root can now change passwords for other user
* timeout if client does not answer fast enough
* Sent LANG as locale to server

Version 1.95
* Make location of ldap config files compile time configurable

Version 1.94
* Add more sanity checks to prevent file corruption

Version 1.93
* Minor bug fixes to rpasswdd

Version 1.92

src/groupdel.c
/* Copyright (C) 2003, 2004, 2005 Thorsten Kukuk
   Author: Thorsten Kukuk <kukuk@suse.de>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2 as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <time.h>
#include <utmp.h>
#include <fcntl.h>
#include <paths.h>
#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <getopt.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <libgen.h>
#include <sys/resource.h>
#ifdef HAVE_LIBNSCD_H
#include <libnscd.h>
#endif

#ifdef USE_LDAP
#include "libldap.h"
#endif

#include "i18n.h"
#include "group.h"
#include "public.h"
#include "logging.h"
#include "utf8conv.h"
#include "logindefs.h"
#include "read-files.h"
#include "error_codes.h"

static void
print_usage (FILE *stream, const char *program)
{
  fprintf (stream, _("Usage: %s [-D binddn] [-P path] group\n"), program);
}

static void
print_help (const char *program)
{
  print_usage (stdout, program);
  fprintf (stdout, _("%s - delete a group\n\n"), program);

#ifdef USE_LDAP
  fputs (_("  -D binddn      Use dn \"binddn\" to bind to the LDAP directory\n"),
         stdout);
#endif
  fputs (_("  -P path        Search passwd, shadow and group file in \"path\"\n"),
	 stdout);
  fputs (_(" --service srv   Add account to nameservice 'srv'\n"), stdout);
  fputs (_("      --help     Give this help list\n"), stdout);
  fputs (_("  -u, --usage    Give a short usage message\n"), stdout);
  fputs (_("  -v, --version  Print program version\n"), stdout);
  fputs (_("Valid services for --service are: files, ldap\n"), stdout);
}

static const char *program = "groupdel";

static struct option long_options[] = {
#ifdef USE_LDAP
  {"binddn",      required_argument, NULL, 'D' },
#endif
  {"force",       no_argument,       NULL, 'f'},
  {"remove-home", no_argument,       NULL, 'r'},
  {"path",        required_argument, NULL, 'P'},
  {"version",     no_argument,       NULL, 'v'},
  {"service",     required_argument, NULL, '\253'},
  {"usage",       no_argument,       NULL, 'u'},
  {"help",        no_argument,       NULL, '\255'},
  {NULL, 0, NULL, '\0'}
};
static const char *short_options = "D:frP:uv";

static struct passwd *
files_getpwent (void)
{
  enum nss_status status;
  static int buflen = 256;
  static char *buffer = NULL;
  static struct passwd resultbuf;

  if (buffer == NULL)
    buffer = malloc (buflen);

  while ((status = files_getpwent_r (&resultbuf, buffer, buflen, &errno))
	 == NSS_STATUS_TRYAGAIN && errno == ERANGE)
    {
      errno = 0;
      buflen += 256;
      buffer = realloc (buffer, buflen);
    }
  if (status == NSS_STATUS_SUCCESS)
    return &resultbuf;
  else
    return NULL;
}

static int
is_primary_group (gid_t gid, int have_extrapath)
{
  struct passwd *pw;
  int retval = 0;

  setpwent ();

  while ((pw = getpwent ()))
    {
      if (pw->pw_gid == gid)
	{
	  fprintf (stderr,
		   _("%s: GID `%u' is primary group of `%s'.\n"),
		   program, (unsigned int) gid,
		   utf8_to_locale (pw->pw_name));
	  sec_log (program, MSG_GID_IS_PRIMARY_GROUP, (unsigned int) gid,
		   pw->pw_name, getuid ());
	  retval = 1;
	}
    }

  endpwent ();

  if (have_extrapath)
    while ((pw = files_getpwent ()))
      {
      if (pw->pw_gid == gid)
	{
	  fprintf (stderr,
		   _("%s: GID `%u' is primary group of `%s'.\n"),
		   program, (unsigned int) gid,
		   utf8_to_locale (pw->pw_name));
	  sec_log (program, MSG_GID_IS_PRIMARY_GROUP, (unsigned int) gid,
		   pw->pw_name, getuid ());
	  retval = 1;
	}
      }

  return retval;
}

int
main (int argc, char **argv)
{
  char *use_service = NULL;
#ifdef USE_LDAP
  char *binddn = NULL;
#endif
  char *remove_group;
  int have_extrapath = 0;
  group_t *gr_data;
  int retval = E_SUCCESS;

#ifdef ENABLE_NLS
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
#endif

  /* Before going any further, raise the ulimit and ignore
     signals.  */

  init_environment ();

  open_sec_log (program);

  while (1)
    {
      int c;
      int option_index = 0;

      c = getopt_long (argc, argv, short_options,
		       long_options, &option_index);
      if (c == (-1))
	break;
      switch (c)
	{
#ifdef USE_LDAP
	case 'D':
	  binddn = optarg;
	  break;
#endif
	case 'P':
	  files_etc_dir = strdup (optarg);
	  have_extrapath = 1;
	  /* If -P option is used, set service to "files" if not already
	     set through an option. If we don't limitate to service files,
	     we can get trouble finding the right source.  */
	  if (!use_service)
	    use_service = "files";
	  break;
	case '\253':
	  if (use_service != NULL)
            {
              print_usage (stderr, program);
              return E_BAD_ARG;
            }

          if (strcasecmp (optarg, "files") == 0)
            use_service = "files";
#ifdef USE_LDAP
          else if (strcasecmp (optarg, "ldap") == 0)
            use_service = "ldap";
#endif
          else
            {
              fprintf (stderr, _("Service `%s' not supported.\n"), optarg);
              print_usage (stderr, program);
              return E_BAD_ARG;
            }
	  break;
	case '\255':
	  print_help (program);
	  return 0;
	case 'u':
	  print_usage (stdout, program);
	  return 0;
	case 'v':
	  print_version (program, "2005");
	  return 0;
	default:
	  print_error (program);
	  return E_USAGE;
	}
    }

  argc -= optind;
  argv += optind;

  if (argc > 1)
    {
      fprintf (stderr, _("%s: Too many arguments.\n"), program);
      print_error (program);
      return E_USAGE;
    }
  else if (argc == 0)
    {
      fprintf (stderr, _("%s: Too few arguments.\n"), program);
      print_error (program);
      return E_USAGE;
    }
  else
    {
      int buflen = 256;
      char *buffer = alloca (buflen);
      struct passwd resultbuf;
      struct passwd *pw;

      /* Determine our own user name for PAM authentication.  */
      while (getpwuid_r (getuid (), &resultbuf, buffer, buflen, &pw) != 0
	     && errno == ERANGE)
	{
	  errno = 0;
	  buflen += 256;
	  buffer = alloca (buflen);
	}

      if (!pw)
	{
	  sec_log (program, MSG_NO_ACCOUNT_FOUND, getuid ());
	  fprintf (stderr, _("%s: Cannot determine your user name.\n"),
		   program);
	  return E_NOTFOUND;
	}

      if (do_authentication ("shadow", pw->pw_name, NULL) != 0)
	{
          sec_log (program, MSG_PERMISSION_DENIED, pw->pw_name,
                   pw->pw_uid, getuid());
          return E_NOPERM;
        }
    }

  remove_group = locale_to_utf8 (argv[0]);

  gr_data = find_group_data (remove_group, 0, use_service);
  if (gr_data == NULL || gr_data->service == S_NONE)
    {
      if (use_service)
	fprintf (stderr, _("%s: Group `%s' not found in service `%s'.\n"),
		 program, utf8_to_locale (remove_group), use_service);
      else
	fprintf (stderr, _("%s: Unknown group `%s'.\n"), program,
		 utf8_to_locale (remove_group));

      sec_log (program, MSG_UNKNOWN_GROUP, remove_group, getuid ());

      return E_NOTFOUND;
    }

  if (is_primary_group (gr_data->gr.gr_gid, have_extrapath))
    {
      fprintf (stderr, _("%s: Cannot remove user's primary group.\n"),
	       program);
      sec_log (program, MSG_CANNOT_REMOVE_PRIMARY_GROUP,
	       gr_data->gr.gr_name, getuid());
      return E_GROUP_BUSY;
    }

#ifdef USE_LDAP
  if (gr_data->service == S_LDAP)
    {
      if (binddn == NULL)
	{
	  binddn = get_caller_dn ();
	  if (binddn == NULL)
	    {
	      fprintf (stderr, _("%s: Cannot delete group from LDAP database without DN.\n"),
		       program);
	    }
	  else gr_data->binddn = strdup (binddn);
	}
      else
	gr_data->binddn = strdup (binddn);

      if (gr_data->oldclearpwd == NULL)
	{
	  char *cp = get_ldap_password (gr_data->binddn);

	  if (cp)
	    gr_data->oldclearpwd = strdup (cp);
	  else
	    {
	      fprintf (stderr,
		       _("%s: Group not deleted from LDAP database.\n"),
		       program);
	      return E_FAILURE;
	    }
	}
    }
#endif

#if 0 /* XXX */
  i = call_script ("GROUPDEL_PRECMD", pw_data->pw.pw_name, pw_data->pw.pw_uid,
		   pw_data->pw.pw_gid, pw_data->pw.pw_dir);
  if (i != 0)
    {
      fprintf (stderr, _("%s: GROUPDEL_PRECMD fails with exit code %d.\n"),
	       program, i);
      return E_FAILURE;
    }
#endif

  /* Lock group file, so that a concurrent processes will not
     use this group.  */
  if (gr_data->service == S_LOCAL && lock_database () != 0)
    {
      fputs (_("Cannot lock group file: already locked.\n"), stderr);
      sec_log (program, MSG_GROUP_FILE_ALREADY_LOCKED);
      return E_PWDBUSY;
    }

  gr_data->todo = DO_DELETE;
  if (write_group_data (gr_data, 1) != 0)
    {
      fprintf (stderr, _("%s: Error deleting group `%s'.\n"),
	       program, utf8_to_locale (gr_data->gr.gr_name));
      sec_log (program, MSG_ERROR_REMOVING_GROUP, gr_data->gr.gr_name,
	       gr_data->gr.gr_gid, getuid ());
      free_group_t (gr_data);
      return E_FAILURE;
    }
  else
    sec_log (program, MSG_GROUP_DELETED, gr_data->gr.gr_name,
	     gr_data->gr.gr_gid, getuid ());

#ifdef HAVE_NSCD_FLUSH_CACHE
  /* flush NSCD cache to remove group really from the system.  */
  nscd_flush_cache ("group");
#endif

  if (gr_data->service == S_LOCAL)
    ulckpwdf ();

#if 0 /* XXX */
  i = call_script ("GROUPDEL_POSTCMD", pw_data->pw.pw_name, pw_data->pw.pw_uid,
		   pw_data->pw.pw_gid, pw_data->pw.pw_dir);
  if (i != 0)
    {
      fprintf (stderr, _("%s: GROUPDEL_POSTCMD fails with exit code %d.\n"),
	       program, i);
      return E_FAILURE;
    }
#endif

  free_group_t (gr_data);

  return retval;
}
